howtos/how-to-make-a-threaded-chat/server.js

```diff
@@ -0,0 +1,228 @@
+const receiverId = `threadsExampleReceiver`;
+
+// Create a thread as a new conversation
+async function createThread(parentMessageId, parentConvId, participants) {
+  const conversationId = "replyto_" + parentMessageId;
+  return fetch(`${basePath}/v1/${appId}/conversations/${conversationId}`, {
+    method: "PUT",
+    headers: {
+      "Content-Type": "application/json",
+      Authorization: `Bearer ${secretKey}`,
+    },
+    body: JSON.stringify({
+      participants: participants,
+      subject: "Replies",
+      custom: {
+        parentConvId: parentConvId,
+        parentMessageId: parentMessageId,
+      },
+    }),
+  });
+}
+
+async function duplicateParentMessageText(parentMessageId, messageText) {
+  const conversationId = "replyto_" + parentMessageId;
+  return fetch(
+    `${basePath}/v1/${appId}/conversations/${conversationId}/messages`,
+    {
+      method: "POST",
+      headers: {
+        "Content-Type": "application/json",
+        Authorization: `Bearer ${secretKey}`,
+      },
+      body: JSON.stringify([
+        {
+          text: messageText,
+          type: "SystemMessage",
+        },
+      ]),
+    }
+  );
+}
+
+async function updateReplyCount(messageId, conversationId, count) {
+  return fetch(
+    `${basePath}/v1/${appId}/conversations/${conversationId}/messages/${messageId}`,
+    {
+      method: "PUT",
+      headers: {
+        "Content-Type": "application/json",
+        Authorization: `Bearer ${secretKey}`,
+      },
+      body: JSON.stringify({
+        custom: { replyCount: count.toString() },
+      }),
+    }
+  );
+}
+
+function getMessages(messageId) {
+  const conversationId = "replyto_" + messageId;
+
+  return fetch(
+    `${basePath}/v1/${appId}/conversations/${conversationId}/messages`,
+    {
+      method: "GET",
+      headers: {
+        "Content-Type": "application/json",
+        Authorization: `Bearer ${secretKey}`,
+      },
+    }
+  );
+}
+
+app.post("/newThread", async (req, res) => {
+  // Get details of the message we'll reply to
+  const parentMessageId = req.body["messageId"];
+  const parentConvId = req.body["conversationId"];
+  const parentMessageText = req.body["messageText"];
+  const parentParticipants = req.body["participants"];
+
+  let response = await getMessages(parentMessageId);
+  let messages = await response.json();
+
+  // Create a message with the text of the parent message if one doesn't already exist
+  if (messages.data === undefined || messages.data.length == 0) {
+    await createThread(parentMessageId, parentConvId, parentParticipants);
+    await duplicateParentMessageText(parentMessageId, parentMessageText);
+  }
+
+  res.status(200).end();
+});
+
+// Endpoint for message.sent webhook
+app.post("/updateReplyCount", async (req, res) => {
+  const data = req.body.data;
+  const conversationId = data.conversation.id;
+  const messageType = data.message.type;
+
+  if (
+    conversationId.slice(0, 8) === "replyto_" &&
```

Comment:
a neat trick might be to use `"replyto_".startsWith("replyto_")` for readability

Reply:
Definitely nicer!